dm-manager/pkg/devices/devices.go

```diff
@@ -0,0 +1,232 @@
+/*
+ * // SPDX-FileCopyrightText: (C) 2025 Intel Corporation
+ * // SPDX-License-Identifier: Apache-2.0
+ */
+
+package devices
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+
+	computev1 "github.com/open-edge-platform/infra-core/inventory/v2/pkg/api/compute/v1"
+	inventoryv1 "github.com/open-edge-platform/infra-core/inventory/v2/pkg/api/inventory/v1"
+	"github.com/open-edge-platform/infra-core/inventory/v2/pkg/client"
+	"github.com/open-edge-platform/infra-core/inventory/v2/pkg/errors"
+	"github.com/open-edge-platform/infra-core/inventory/v2/pkg/logging"
+	"github.com/open-edge-platform/infra-external/dm-manager/pkg/api/mps"
+	"github.com/open-edge-platform/infra-external/dm-manager/pkg/api/rps"
+	rec_v2 "github.com/open-edge-platform/orch-library/go/pkg/controller/v2"
+)
+
+const (
+	powerOn    = mps.PowerActionRequestActionN2
+	powerOff   = mps.PowerActionRequestActionN8
+	powerReset = mps.PowerActionRequestActionN10
+	// In-band power actions require the presence of an agent running while the operating system
+	// is up and operational like the Intel Local Manageability Service (LMS).
+	powerSleep     = mps.PowerActionRequestActionN4
+	powerHibernate = mps.PowerActionRequestActionN7
+)
+
+var log = logging.GetLogger("DeviceReconciler")
+
+var powerMapping = map[computev1.PowerState]mps.PowerActionRequestAction{
+	computev1.PowerState_POWER_STATE_UNSPECIFIED: powerOn, // todo: consider removing this mapping
+	computev1.PowerState_POWER_STATE_ON:          powerOn,
+	computev1.PowerState_POWER_STATE_OFF:         powerOff,
+	computev1.PowerState_POWER_STATE_SLEEP:       powerSleep,
+	computev1.PowerState_POWER_STATE_RESET:       powerReset,
+	computev1.PowerState_POWER_STATE_HIBERNATE:   powerHibernate,
+}
+
+type DeviceID string
+
+func (id DeviceID) GetTenantID() string {
+	return strings.Split(string(id), "_")[0]
+}
+
+func (id DeviceID) GetHostUUID() string {
+	return strings.Split(string(id), "_")[1]
+}
+
+func (id DeviceID) String() string {
+	return fmt.Sprintf("[tenantID=%s, hostID=%s]", id.GetTenantID(), id.GetHostUUID())
+}
+
+func NewDeviceID(tenantID, hostUUID string) DeviceID {
+	return DeviceID(tenantID + "_" + hostUUID)
+}
+
+type DeviceController struct {
+	MpsClient          mps.ClientWithResponsesInterface
+	RpsClient          rps.ClientWithResponsesInterface
+	InventoryRmClient  client.TenantAwareInventoryClient // manages Current* fields
+	InventoryAPIClient client.TenantAwareInventoryClient // manages Desired* fields
```

Comment:
you are not supposed to have API client. As RM you can only touch CurrentFields - if exceptions are required we should make it in OPA